lisa/energy_model.py

```diff
@@ -758,107 +764,163 @@ def get_optimal_placements(self, capacities, capacity_margin_pct=0):
         return ret
 
     @classmethod
-    def _find_core_groups(cls, target):
+    def check_from_target(cls, target):
         """
-        Read the core_siblings masks for each CPU from sysfs
+        Check if an :class:`EnergyModel` can be loaded from the target.
 
-        :param target: Devlib Target object to read masks from
-        :returns: A list of tuples of ints, representing the partition of core
-                  siblings
+        :param target: Target to look at.
+        :type target: devlib.target.Target
         """
-        cpus = list(range(target.number_of_cpus))
+        try:
+            cls._find_subcls(target)
+        except TargetStableError:
+            return False
+        else:
+            return True
 
-        topology_base = '/sys/devices/system/cpu/'
+    @classmethod
+    def _find_subcls(cls, target):
+        subcls_list = sorted(
+            get_subclasses(cls, only_leaves=True),
+            key=lambda cls: cls._AUTODETECT_ORDER
+        )
 
-        # We only care about core_siblings, but let's check *_siblings, so we
-        # can throw an error if a CPU's thread_siblings isn't just itself, or if
-        # there's a topology level we don't understand.
+        for subcls in subcls_list:
+            if subcls.check_from_target(target):
+                return subcls
 
-        # Since we might have to read a lot of files, read everything we need in
-        # one go to avoid taking too long.
-        mask_glob = topology_base + 'cpu**/topology/*_siblings'
-        file_values = read_multiple_oneline_files(target, [mask_glob])
+        raise TargetStableError('Unable to probe for energy model on target.')
 
-        regex = re.compile(
-            topology_base + r'cpu([0-9]+)/topology/([a-z]+)_siblings')
+    @classmethod
+    def from_target(cls, target):
+        """
+        Create an instance of (a subclass of) :class:``EnergyModel`` by reading a target filesystem.
 
-        ret = set()
+        If present, load an EM provided via dt using from_sd_target, since these
+        devices make the full EM available via the sched domain in sysfs. If
+        there is no EM at this location, attempt to load the simplified EM
+        made available via dedicated sysfs files.
```

Comment:
I see this was an existing comment, but it is broken no ? We seem to look at the simplified EM first (according to `_AUTODETECT_ORDER`), and _then_ to the full/legacy EM. Or maybe I got that wrong ?

Also, I think we should say the simple EM is exposed in _debugfs_, and the legacy EM in _procfs_, rather than sysfs.

Reply:
Yes, I replaced it by a .. seealso pointing at the others from_target()